In [ ]:
import pandas as pd

ml_table = pd.read_csv("../data/processed/ml_table_raw.csv")
print(ml_table.shape)

(99441, 17)


In [2]:
# تحويل الأعمدة من نص (string) لتاريخ حقيقي (datetime) عشان نقدر نقارنهم
ml_table["order_delivered_customer_date"] = pd.to_datetime(ml_table["order_delivered_customer_date"])
ml_table["order_estimated_delivery_date"] = pd.to_datetime(ml_table["order_estimated_delivery_date"])

# بناء الـ label: 1 لو الطلب متأخر، 0 لو وصل بالوقت أو قبله
ml_table["is_late"] = (ml_table["order_delivered_customer_date"] > ml_table["order_estimated_delivery_date"]).astype(int)

print(ml_table["is_late"].value_counts())

is_late
0    91614
1     7827
Name: count, dtype: int64


### Class Imbalance Check

- Total orders: 99,441
- On-time (0): 91,614 (~92.13%)
- Late (1): 7,827 (~7.87%)

**Conclusion:** Clear class imbalance problem — late orders represent 
only ~7.87% of the data. Accuracy alone would be misleading (a model 
predicting "always on-time" would score ~92% without learning anything).
This will be addressed in Notebook 6 by choosing an appropriate metric 
(e.g., precision/recall, F1, or ROC-AUC) instead of relying on accuracy alone.

In [3]:
check_sample = ml_table[["order_id", "order_delivered_customer_date", 
                           "order_estimated_delivery_date", "is_late"]].sample(5, random_state=42)
print(check_sample)

                               order_id order_delivered_customer_date  \
52263  b9a6c5f5df52c7226ac85aee7524c27f           2018-06-19 12:44:08   
46645  261e71d2349c713eafa9f3df5972b95d           2018-01-30 11:32:35   
37546  67b50899f52995848c427e361e10dde3           2018-06-27 13:17:27   
94756  32733fc014b67ef70fa6039dd8c6ba82           2017-09-25 17:53:23   
14771  39a70e9e9b729b11dee34ac12478597f           2017-08-22 16:45:00   

      order_estimated_delivery_date  is_late  
52263                    2018-07-17        0  
46645                    2018-02-15        0  
37546                    2018-07-16        0  
94756                    2017-09-22        1  
14771                    2017-09-12        0  


In [ ]:
ml_table.to_csv("../data/processed/ml_table_labeled.csv", index=False)
print("Saved successfully!")

Saved successfully!
